### PR TITLE
Set comma-dangle to 'always-multiline'

### PR DESCRIPTION
**Pull request by bluefuton**
Require a trailing comma for multiline object literals (but not single line).

Docs here:

http://eslint.org/docs/rules/comma-dangle

Good writeup on the rationale here:

https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8

### CODE REVIEW

**Comment by aduth** [comment]
I don't know why you'd choose to assign me as reviewer knowing how at odds I am with this rule 😆 

Reiterating a point I made elsewhere that received no rebuttal, I think people like comma dangle for the simple one-member-per-line cases but aren't so inclined to it on the multi-line cases.

Let's take your recent commit https://github.com/Automattic/wp-calypso/commit/5ebd482c8cb237b4df6edb5202eeac6c1972e02d for example. (Sorry, not trying to pick on you!)

Here you correctly followed `comma-dangle` per the rule and I assume your preference:

https://github.com/Automattic/wp-calypso/commit/5ebd482#diff-025e987e50d0be50cc8945d08938c0f9R41

But the very next lines you'd not found it natural to adhere to the rule where it should also have applied, omitting commas at the close of two multi-line objects:

https://github.com/Automattic/wp-calypso/commit/5ebd482#diff-025e987e50d0be50cc8945d08938c0f9R42
https://github.com/Automattic/wp-calypso/commit/5ebd482#diff-025e987e50d0be50cc8945d08938c0f9R43

I seem to be in the minority with my views toward this rule, so I'll not block it being introduced, but I would very much like to see it enforced consistently.

**Comment by bluefuton** [comment]
I haven't really got a really strong preference for using this style in particular, but like you, I would like us to adopt a standard and stick to that. Even if it means me going back and fixing my recent work :D

We've had a few quibbles about it on Reader in code reviews, and it'd be good to say "that's the way we do it in Calypso", whichever way we decide.

I notice Airbnb (whose eslint guidelines seem a bit of a benchmark) use `always-multiline` across the board:

https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/errors.js#L4

**Comment by blowery** [comment]
ship it.

**Comment by blowery** [comment]
To be honest, I'm not in love with this rule either (too many years lost to IE dying trying to parse trailing commas), but I'm warming up to it. Consistent enforcement is more important than my preferences.

**Comment by nb** [comment]
I love this rule and I think it has been worth following it over the years, even taking into account the fraction of a youth lost to the old IE comma crisis.

**Comment by aduth** [comment]
Revisiting this, we've been [using this rule in the Gutenberg project](https://github.com/WordPress/gutenberg/blob/c521550e53470b791eed2a022f47ee10fc361b23/.eslintrc.json#L40) for about a month now. The large/complex object dangle is still a bit awkward, but it's not been as painful as I might have originally assumed.

Another interesting case that's easy to overlook is function arguments:

```js
doThing(
	'foo',
	'bar',
	'baz' // To dangle, or not to dangle?
);
```

The answer depends on which version of the spec you're following. [Trailing commas on function arguments](https://github.com/tc39/proposal-trailing-function-commas) was [ratified for ES2017](https://github.com/tc39/proposals/blob/master/finished-proposals.md). This is supported by ESLint's `comma-dangle` rule but is configured to `ignore` by default.

http://eslint.org/docs/rules/comma-dangle#options

**Comment by deBhal** [comment]
TL;DR: We should add a changelog entry and 🚢 .

Right now we specify `ecmaVersion: 7` (2016), so we should probably hold off on the function arguments case for now, and revisit once we bump the `ecmaVersion` to a spec that makes it legal syntax :)

Turns out that the rule will won't enable arguments unless you explicitly specify them ([comma-dangle.js](https://github.com/eslint/eslint/blob/master/lib/rules/comma-dangle.js#L56)), so I think the rule is right as-is.

> Reiterating a point I made elsewhere that received no rebuttal, I think people like comma dangle for the simple one-member-per-line cases but aren't so inclined to it on the multi-line cases.

Quick rebuttal: I think your examples actually highlight the value of enforcing the rule. People are definitely inconsistent about adding the commas in the multi-line cases (I know I am :( ), but I bet you'd get a very solid consensus if you ask people which of these diffs they prefer:

calypso_ _less_ _git_diff_ _69x37_and_calypso_ _less_ _git_diff_head _ _70x88

**Comment by mcsf** [comment]
:shipit: :shipit: :shipit: , so we can npm-publish the latest additions (#5, #8, #9)

**Comment by mcsf** [comment]
@bluefuton, what's the status here?

**Comment by bluefuton** [comment]
@mcsf looks like we need a changelog entry (as mentioned above https://github.com/Automattic/eslint-config-wpcalypso/pull/5#issuecomment-308898273) and presumably to bump the version to v0.9.1? I'm not familiar with the release process.

cc @aduth

**Comment by aduth** [comment]
This rule is included in `eslint-config-wpcalypso@1.0.0`
